source/time_tables/previous_passages.cpp

```diff
@@ -0,0 +1,163 @@
+                                  navitia::to_posix_time(handler.max_datetime, data));
+
+    for(auto dt_stop_time : passages_dt_st) {
+        pbnavitia::Passage * passage;
+        if(vis.api_str == "NEXT_ARRIVALS")
+            passage = handler.pb_response.add_next_arrivals();
+        else
+            passage = handler.pb_response.add_next_departures();
+        auto departure_date = navitia::to_posix_timestamp(dt_stop_time.first, data);
+        auto arrival_date = navitia::to_posix_timestamp(dt_stop_time.first, data);
+        passage->mutable_stop_date_time()->set_departure_date_time(departure_date);
+        passage->mutable_stop_date_time()->set_arrival_date_time(arrival_date);
+        const type::JourneyPatternPoint* jpp = dt_stop_time.second->journey_pattern_point;
+        fill_pb_object(jpp->stop_point, data, passage->mutable_stop_point(),
+                depth, now, action_period);
+        const type::VehicleJourney* vj = dt_stop_time.second->vehicle_journey;
+        const type::JourneyPattern* jp = vj->journey_pattern;
+        const type::Route* route = jp->route;
+        const type::Line* line = route->line;
+        const type::PhysicalMode* physical_mode = jp->physical_mode;
+        auto m_vj = passage->mutable_vehicle_journey();
+        auto m_route = m_vj->mutable_route();
+        auto m_physical_mode = m_vj->mutable_journey_pattern()->mutable_physical_mode();
+        fill_pb_object(vj, data, m_vj, 0, now, action_period, show_codes);
+        fill_pb_object(route, data, m_route, 0, now, action_period, show_codes);
+        fill_pb_object(line, data, m_route->mutable_line(), 0, now, action_period, show_codes);
+        fill_pb_object(physical_mode, data, m_physical_mode, 0, now, action_period);
+    }
+    auto pagination = handler.pb_response.mutable_pagination();
+    pagination->set_totalresult(total_result);
+    pagination->set_startpage(start_page);
+    pagination->set_itemsperpage(count);
+    pagination->set_itemsonpage(passages_dt_st.size());
+    return handler.pb_response;
+}
+
+
+pbnavitia::Response previous_departures(const std::string &request,
+        const std::vector<std::string>& forbidden_uris,
+        const pt::ptime datetime, uint32_t duration, uint32_t nb_stoptimes,
+        const int depth, const type::AccessibiliteParams & accessibilite_params,
+        const type::Data & data, bool disruption_active, uint32_t count, uint32_t start_page,
+        const bool show_codes) {
+
+    struct vis_previous_departures {
+        struct predicate_t {
+            const type::Data &data;
+            predicate_t(const type::Data& data) : data(data){}
+            bool operator()(const type::idx_t jppidx) const {
+                auto jpp = data.pt_data->journey_pattern_points[jppidx];
+                auto last_jpp = jpp->journey_pattern->journey_pattern_point_list.back();
+                return jpp == last_jpp;
+            }
```

Comment:
I think you want to check if this is a valid pick_up, not only remove the last jpp.